pop warning message for multiprocessing module when asan is built in

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35098 pop warning message for multiprocessing module when asan is built in**

Pytorch multiprocessing module (and standard multiprocessing module as well) is not compatible with asan built when using spawn-based multiprocessing. Adding a warning message so people who tends to use spawn-based multiprocessing will know that they need a non-asan built.

Here is an example shows the warning msg and the error when using spawn-based multiprocessing:
```
In [1]: import torch.multiprocessing as mp                                                                                                                                                                                        
Warning: torch.multiprocessing: your pytorch binary has address sanitizer (asan) built in, asan is currently not compatible with spawn-based (start method is 'spawn') multiprocessing which is provided in this module, you might get unexpected behavior (eg. missing attribute, crash, etc.), please rebuild pytorch without asan if you need spawn-based multiprocessing (multiprocessing_init at caffe2/torch/csrc/multiprocessing/init.cpp:28)

In [2]: def foo(): 
   ...:     print("foo is running") 
   ...:                                                                                                                                                                                                                           

In [3]: ctx = mp.get_context('fork')                                                                                                                                                                                              

In [4]: p = ctx.Process(target=foo)                                                                                                                                                                                               

In [5]: p.start()                                                                                                                                                                                                                 

foo is running
In [6]: p.join()                                                                                                                                                                                                                  

In [7]: ctx = mp.get_context('spawn')                                                                                                                                                                                             

In [8]: p = ctx.Process(target=foo)                                                                                                                                                                                              

In [9]: p.start()                                                                                                                                                                                                                

In [10]: Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/fbcode/platform007/lib/python3.7/multiprocessing/spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "/usr/local/fbcode/platform007/lib/python3.7/multiprocessing/spawn.py", line 115, in _main                                                                                                                                 
    self = reduction.pickle.load(from_parent)
AttributeError: Can't get attribute 'foo' on <module '__main__' (built-in)>
In [11]:                                                                      
```
Differential Revision: [D20558202](https://our.internmc.facebook.com/intern/diff/D20558202)

Commit:
pop warning message for multiprocessing module when asan is built in

[ghstack-poisoned]

glaringlee commented:
asan is not compatible with both torch.multiprocessing and standard python multiprocessing module. 
This is the best place I can think to give warning for both module.

gchanan commented on torch/csrc/multiprocessing/init.cpp:
```diff
@@ -19,6 +19,15 @@ namespace multiprocessing {
+#if __has_feature(address_sanitizer)
+  TORCH_WARN(
+    "torch.multiprocessing: your pytorch binary has address sanitizer (asan) built in, "
+    "asan is currently not compatiable with spawn-based (start method is 'spawn') multiprocessing "
```
nit: compatible

glaringlee replied:
updated

gchanan commented:
Echoing a chat I had with @colesbury:

Can you print the behavior of a program that fails in fbcode (and in OSS) where the warning comes up? Might be worth considering making this an error if it always causes an error later anyway.

Commit:
Update on "pop warning message for multiprocessing module when asan is built in"


Differential Revision: [D20558202](https://our.internmc.facebook.com/intern/diff/D20558202)

[ghstack-poisoned]

glaringlee commented:
Added an example showing when the warning message will be popped up and what kind of error asan will cause in multiprocessing module. Since 'fork' method still works and it is default in most of the platforms (only windows use spawn as default), a warning message is enough here I think.
